### PR TITLE
Add unrolling dimension into GPU data tiling

### PR DESCRIPTION
**Pull request by lialan**
TODO: Interleaving dimensions. But that is just changing transpose dimensions.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp [hunk 1]
```diff
@@ -100,6 +102,40 @@ static RankedTensorType transposeIfNarrowNResult(RankedTensorType tensorType) {
   return RankedTensorType::get(newShape, elemType, newEncoding);
 }
 
+template <typename T>
+SmallVector<T> getTileSwizzlingDims(ArrayRef<T> typeShape,
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp]: Are you going to use this refactored function for below two `resolve*` implementations?

File: compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp [hunk 2]
```diff
@@ -112,20 +148,28 @@ resolveTileSwizzlingType(RankedTensorType packedType,
...
-  auto iT2 = intrinsicVectorShape[1];
-  auto oT1 = encodingInfo.innerTileSizes[0] / iT1;
-  auto oT2 = encodingInfo.innerTileSizes[1] / iT2;
+  auto packedShape = packedType.getShape();
+  auto packedRank = packedType.getRank();
+  auto AB1 = packedShape[packedRank - 2] / encodingInfo.innerTileSizes[0];
+  auto AB2 = intrinsicVectorShape[0];
+  auto ABunroll = encodingInfo.innerTileSizes[0] / AB2;
+  auto ACC1 = packedShape[packedRank - 1] / intrinsicVectorShape[1];
+  auto ACC2 = intrinsicVectorShape[1];
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp]: nit: we follow camelCase naming style in IREE.

```suggestion
  auto ab1 = packedShape[packedRank - 2] / encodingInfo.innerTileSizes[0];
  auto ab2 = intrinsicVectorShape[0];
  auto abUnroll = encodingInfo.innerTileSizes[0] / ab2;
  auto acc1 = packedShape[packedRank - 1] / intrinsicVectorShape[1];
  auto acc2 = intrinsicVectorShape[1];
```

**Review by hanhanW — COMMENTED** [review]
I think it is easier if you can add unrolling K + interleaving K to the PR. Because we are now on the same page about how it should look like.

For the next step, I'd suggest to add the intrinsic support for `mfma_i32_32x32x16_i8`; see how we can improve the current infra. The intrinsic vector shape becomes very interesting. They are no longer a 1-D vector. E.g.,

```
./matrix_calculator.py --architecture cdna3 --instruction v_mfma_i32_32x32x16_i8 --matrix-layout --D-matrix
+--------+----------+----------+----------+----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+
|   lane | v0       | v1       | v2       | v3       | v4        | v5        | v6        | v7        | v8        | v9        | v10       | v11       | v12       | v13       | v14       | v15       |
+========+==========+==========+==========+==========+===========+===========+===========+===========+===========+===========+===========+===========+===========+===========+===========+===========+
|      0 | D[0][0]  | D[1][0]  | D[2][0]  | D[3][0]  | D[8][0]   | D[9][0]   | D[10][0]  | D[11][0]  | D[16][0]  | D[17][0]  | D[18][0]  | D[19][0]  | D[24][0]  | D[25][0]  | D[26][0]  | D[27][0]  |
+--------+----------+----------+----------+----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+
```

File: compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp (hunk 1 above)
Comment (lialan) [reply compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp]: yup, combined everything together

**Comment by hanhanW** [comment]
> I think it is easier if you can add unrolling K + interleaving K to the PR. Because we are now on the same page about how it should look like.
> 
> For the next step, I'd suggest to add the intrinsic support for `mfma_i32_32x32x16_i8`; see how we can improve the current infra. The intrinsic vector shape becomes very interesting. They are no longer a 1-D vector. E.g.,
> 
> ```
> ./matrix_calculator.py --architecture cdna3 --instruction v_mfma_i32_32x32x16_i8 --matrix-layout --D-matrix
> +--------+----------+----------+----------+----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+
> |   lane | v0       | v1       | v2       | v3       | v4        | v5        | v6        | v7        | v8        | v9        | v10       | v11       | v12       | v13       | v14       | v15       |
> +========+==========+==========+==========+==========+===========+===========+===========+===========+===========+===========+===========+===========+===========+===========+===========+===========+
> |      0 | D[0][0]  | D[1][0]  | D[2][0]  | D[3][0]  | D[8][0]   | D[9][0]   | D[10][0]  | D[11][0]  | D[16][0]  | D[17][0]  | D[18][0]  | D[19][0]  | D[24][0]  | D[25][0]  | D[26][0]  | D[27][0]  |
> +--------+----------+----------+----------+----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+
> ```

The comment is not related to the unrolling support. It is more about adding supports for other intrinsics.

I just found that I pointed out wrong intrinsic. For the (i8, i8, i32) intrinsic, we can focus on `mfma_i32_16x16x32_i8` for now. I'll write things down in a doc. TLDR is that we can start adding support for `(f16, f16, f32)` and `(i8, i8, i32)` intrinsic.

According to Benoit's writeup, the `v_mfma_i32_32x32x16_i8` intrinsic is thought to be more power-efficient but lower peak; it will likely enter the picture as a later refinement once we have basic things working.

**Review by Max191 — CHANGES_REQUESTED** [review]
Let's revisit this once @bjacob has simplified the encoding info. I think this logic may no longer be needed.

**Pull request closed without merging**